typing/typedecl.ml

```diff
@@ -395,9 +406,12 @@ let transl_declaration env sdecl (id, uid) =
       Option.is_none unboxed_attr
     | _ -> false, false (* Not unboxable, mark as boxed *)
   in
+  verify_unboxed_attr unboxed_attr sdecl;
   let (tkind, kind) =
     match sdecl.ptype_kind with
-      | Ptype_abstract -> Ttype_abstract, Type_abstract
+      | Ptype_abstract ->
+        let immediate = Type_immediacy.of_attributes sdecl.ptype_attributes in
```

Comment:
Instead of just looking at attributes, would it make sense to also look at the manifest? That is, I think it's better to have `type t = int` be remembered to be an immediate, whether or not the user wrote `[@@immediate]`.

Reply:
At the moment, the idea is that we record here the presence of an annotation for use by `check_coherence` later (which confirms that if the annotation appeared, the thing is actually immediate).

We could certainly improve the kind, though not very well at this point in the code because `transl_declaration` doesn't have access to the manifests of mutually defined types (so your example would work, but not `type t = s and s = int`).  In the layouts prototype, we do something very much like that after building the final environment.  But I'd suggest leaving that as a future improvement, mainly because I'd like for this not to diverge much from the upstream version to make our lives easier in future rebases.

Reply:
Ah, right. There was a little voice in my head warning about mutually recursive types when I wrote my comment, but I ignored it. OK.

The goal of this patch is to make the code generation of the layouts patch the same as pre-layouts. So I'd be worried if the layouts patch improved on this. But I'm pretty sure the improvement is just an optimization to compile times, allowing less expansion. It shouldn't affect the actual result of an immediacy check. So I agree with your choice here.

Reply:
(I agree there is a chance we'll have to come back and improve this to do the layouts test.)